modules/node/src/appRegistry/appRegistry.service.ts

```diff
@@ -41,25 +43,103 @@ export class AppRegistryService {
-      return registryAppInfo;
-    }
-    await this.cfCoreService.installApp(data.data.appInstanceId);
-    return registryAppInfo;
+  async validateAndInstallOrReject(
+    appInstanceId: string,
+    proposeInstallParams: CFCoreTypes.ProposeInstallParams,
+    from: string,
+  ): Promise<void> {
+    let registryAppInfo: AppRegistry;
+    try {
+      registryAppInfo = await this.verifyAppProposal(proposeInstallParams, from);
+      if (registryAppInfo.name !== CoinBalanceRefundApp) {
+        await this.cfCoreService.installApp(appInstanceId);
+      } else {
+        logger.log(`Not installing coin balance refund app, returning registry information`);
+      }
+    } catch (e) {
+      if (!e.message.includes(`Node has insufficient balance`)) {
+        logger.warn(`App install failed, . Error: ${e.stack || e.message}`);
+        await this.cfCoreService.rejectInstallApp(appInstanceId);
+        return;
+      }
+      // try to collateralize, and re-install
+      const tx = await this.channelService.rebalance(
+        from,
+        proposeInstallParams.responderDepositTokenAddress,
+        RebalanceType.COLLATERALIZE,
+        bigNumberify(proposeInstallParams.responderDeposit),
+      );
+      if (tx) {
+        await tx.wait();
+      }
+      try {
+        await this.cfCoreService.installApp(appInstanceId);
+      } catch (e) {
+        logger.warn(`App install failed, . Error: ${e.stack || e.message}`);
+        await this.cfCoreService.rejectInstallApp(appInstanceId);
+      }
+    }
+    await this.runPostInstallTasks(registryAppInfo, appInstanceId, proposeInstallParams, from);
+  }
+
+  private async runPostInstallTasks(
+    registryAppInfo: AppRegistry,
+    appInstanceId: string,
+    proposeInstallParams: CFCoreTypes.ProposeInstallParams,
+    from: string,
+  ) {
+    switch (registryAppInfo.name) {
+      case SimpleLinkedTransferApp:
+        logger.debug(`Saving linked transfer`);
+        const initialState = proposeInstallParams.initialState as SimpleLinkedTransferAppStateBigNumber;
+
+        const isResolving = proposeInstallParams.responderDeposit.gt(Zero);
+        if (isResolving) {
+          const transfer = await this.transferService.getLinkedTransferByPaymentId(initialState.paymentId);
+          transfer.receiverAppInstanceId = appInstanceId;
+          await this.linkedTransferRepository.save(transfer);
+          logger.debug(`Updated transfer with receiver appId!`);
+          return;
+        }
+        await this.transferService.saveLinkedTransfer(
+          from,
+          proposeInstallParams.initiatorDepositTokenAddress,
+          bigNumberify(proposeInstallParams.initiatorDeposit),
+          appInstanceId,
+          initialState.linkedHash,
+          initialState.paymentId,
+          proposeInstallParams.meta,
+        );
+        logger.debug(`Linked transfer saved!`);
+        break;
+      // TODO: add something for swap app? maybe for history preserving reasons.
```

Comment:
we could store swap rates used and provided?

Reply:
Interesting idea... I'll write it up as a separate issue.